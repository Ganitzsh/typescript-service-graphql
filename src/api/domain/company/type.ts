import { ResourceID } from '../../common/id';

interface Address {
  addressLine1: string;
  addressLine2?: string;
  country: string;
  city: string;
  zipCode: string;
}

export interface Company {
  id: ResourceID;
  name: string;
  address: Address;
  taxNumber: string;
}
