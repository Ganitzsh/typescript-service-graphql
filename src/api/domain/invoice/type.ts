import { ResourceID } from '../../common/id';
import { Modifier } from '../modifier/type';

export interface Invoice {
  id: string;

  phases: ResourceID[];
  modifier?: Modifier;

  // This value is computed when phases are modified
  // and not at runtime.
  subtotal: number;
  // This value is computed when phases are modified
  // and not at runtime.
  total: number;
  currency: string;

  issuer: ResourceID;
  recipient: ResourceID;

  dueDate?: Date;
  finalized: boolean;
}
