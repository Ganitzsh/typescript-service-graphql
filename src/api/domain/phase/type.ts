import { ResourceID } from '../../common/id';
import { Invoice } from '../invoice/type';
import { Modifier } from '../modifier/type';

export interface TaxRate {
  label: string;
  value: number;
}

export enum ItemType {
  PerHour,
  Quantity,
}

// Depending on the the importance of
// items in the domain, this might need its
// own service and logic. For now, it's fine
// to make it a sub type of Phase.
export interface Item {
  type: ItemType;
  name: string;
  description: string;
  rate: number;
  taxRate: TaxRate;
}

export interface CostItem {
  item: Item;
  quantity: number;
  subtotal: number;
  total: number;
}

export interface Phase {
  id: ResourceID;

  items: CostItem[];
  modifier?: Modifier;

  // This value is computed when the cost items are modified
  // and not at runtime.
  subtotal: number;
  // This value is computed when the cost items are modified
  // and not at runtime.
  total: number;

  invoice?: ResourceID | Invoice;
}
