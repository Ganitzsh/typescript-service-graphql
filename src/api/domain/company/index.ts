import { Company } from './type';

export { Company };
