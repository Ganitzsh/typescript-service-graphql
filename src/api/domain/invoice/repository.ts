import { Context } from '../../common/context';
import { ResourceID } from '../../common/id';
import { Page } from '../../common/page';

import { Invoice } from './type';

export type InvoicesPage = Page<Invoice>;

export interface GetInvoicesFilter {
  page?: {
    limit?: number;
    cursor?: string;
  };
}

export interface InvoiceRepository {
  find(context: Context, filters: GetInvoicesFilter): Promise<InvoicesPage>;
  findById(context: Context, id: ResourceID): Promise<Invoice>;
}
