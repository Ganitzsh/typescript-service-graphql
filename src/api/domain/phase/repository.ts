import { Context } from '../../common/context';
import { ResourceID } from '../../common/id';
import { Phase } from './type';

export const CompanyNotFound = new Error('company not found');

export interface PhaseRepository {
  findByInvoiceId(context: Context, invoiceId: ResourceID): Promise<Phase[]>;
}
