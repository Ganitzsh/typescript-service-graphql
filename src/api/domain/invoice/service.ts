import { Context } from '../../common/context';
import { ResourceID } from '../../common/id';
import { Company } from '../company';
import { CompanyRepository } from '../company/repository';
import { InvoiceRepository, InvoicesPage } from './repository';

import { Invoice } from './type';

export const listInvoices = async (
  context: Context,
  repository: InvoiceRepository,
  page?: {
    limit?: number;
    cursor?: ResourceID;
  },
): Promise<InvoicesPage> => {
  return repository.find(context, {
    page,
  });
};

export const retrieveInvoice = async (
  context: Context,
  repository: InvoiceRepository,
  id: ResourceID,
): Promise<Invoice> => {
  return repository.findById(context, id);
};

export const retrieveInvoiceParties = async (
  context: Context,
  invoiceRepository: InvoiceRepository,
  companyRepository: CompanyRepository,
  invoiceId: ResourceID,
): Promise<{
  issuer: Company;
  recipient: Company;
}> => {
  const invoice = await invoiceRepository.findById(context, invoiceId);

  const [issuer, recipient] = await Promise.all([
    companyRepository.findById(context, invoice.issuer),
    companyRepository.findById(context, invoice.recipient),
  ]);

  return {
    issuer,
    recipient,
  };
};
