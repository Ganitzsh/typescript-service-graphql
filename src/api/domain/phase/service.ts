import { Context } from '../../common/context';
import { ResourceID } from '../../common/id';
import { PhaseRepository } from './repository';

import { Phase } from './type';

export const retrievePhasesForInvoice = async (
  context: Context,
  repository: PhaseRepository,
  invoiceId: ResourceID,
): Promise<Phase[]> => {
  return repository.findByInvoiceId(context, invoiceId);
};
