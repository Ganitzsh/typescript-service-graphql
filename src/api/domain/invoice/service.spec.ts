import { Context } from '../../common/context';
import { Company } from '../company';
import { CompanyRepository } from '../company/repository';

import { InvoiceRepository } from './repository';
import * as invoiceService from './service';
import { Invoice } from './type';

beforeEach((): void => {
  jest.clearAllMocks();
});

const makeMockInvoice = (overrides?: Partial<Invoice>): Invoice => ({
  id: '1',
  phases: ['1', '2'],
  issuer: '1',
  recipient: '2',
  subtotal: 21.0,
  total: 42.0,
  currency: 'EUR',
  finalized: true,
  ...overrides,
});

const makeMockCompany = (overrides?: Partial<Company>): Company => ({
  id: '1',
  name: 'test company',
  address: {
    addressLine1: 'test address line 1',
    addressLine2: 'test address line 2',
    country: 'test country',
    city: 'test city',
    zipCode: 'test zip code',
  },
  taxNumber: 'test tax id',
  ...overrides,
});

const makeMockContext = (): Context => ({
  traceId: '123',
});

const makeMockInvoiceRepository: () => InvoiceRepository =
  (): InvoiceRepository => ({
    find: jest.fn(),
    findById: jest.fn(),
  });

const makeMockCompanyRepository: () => CompanyRepository =
  (): CompanyRepository => ({
    findById: jest.fn(),
  });

describe('invoice service', (): void => {
  describe('listInvoices', (): void => {
    describe('error handling', (): void => {
      it('throws an error when the repository throws an error', async (): Promise<void> => {
        expect.assertions(1);

        const mockContext = makeMockContext();
        const invoiceRepository = makeMockInvoiceRepository();
        const errorToThrow = new Error('test error');

        (
          invoiceRepository.find as jest.MockedFunction<any>
        ).mockRejectedValueOnce(errorToThrow);

        try {
          await invoiceService.listInvoices(mockContext, invoiceRepository);
        } catch (error: unknown) {
          expect(error).toEqual(errorToThrow);
        }
      });
    });

    describe('data persistency', (): void => {
      it('does not apply any paging when page is undefined', async (): Promise<void> => {
        expect.assertions(1);

        const mockContext = makeMockContext();
        const invoiceRepository = makeMockInvoiceRepository();

        await invoiceService.listInvoices(mockContext, invoiceRepository);

        expect(
          invoiceRepository.find as jest.MockedFunction<any>,
        ).toHaveBeenCalledWith(mockContext, { page: undefined });
      });

      it('applies paging when page is set', async (): Promise<void> => {
        expect.assertions(1);

        const mockContext = makeMockContext();
        const invoiceRepository = makeMockInvoiceRepository();

        await invoiceService.listInvoices(mockContext, invoiceRepository, {
          limit: 1,
        });

        expect(
          invoiceRepository.find as jest.MockedFunction<any>,
        ).toHaveBeenCalledWith(mockContext, {
          page: { limit: 1 },
        });
      });

      it('returns a page of invoices', async (): Promise<void> => {
        expect.assertions(1);

        const mockContext = makeMockContext();
        const invoiceRepository = makeMockInvoiceRepository();

        const invoices = [
          makeMockInvoice(),
          makeMockInvoice({ id: '2' }),
          makeMockInvoice({ id: '3' }),
        ];

        (invoiceRepository.find as jest.MockedFunction<any>).mockResolvedValue({
          data: invoices,
          total: 3,
          cursor: undefined,
        });

        const page = await invoiceService.listInvoices(
          mockContext,
          invoiceRepository,
        );

        expect(page).toMatchSnapshot();
      });
    });
  });

  describe('retrieveInvoice', (): void => {
    describe('error handling', (): void => {
      it('throws an error when the repository throws an error', async (): Promise<void> => {
        expect.assertions(1);

        const invoiceRepository = makeMockInvoiceRepository();
        const errorToThrow = new Error('test error');

        (
          invoiceRepository.findById as jest.MockedFunction<any>
        ).mockRejectedValueOnce(errorToThrow);

        try {
          await invoiceService.retrieveInvoice(
            makeMockContext(),
            invoiceRepository,
            '123',
          );
        } catch (error: unknown) {
          expect(error).toEqual(errorToThrow);
        }
      });
    });

    describe('data persistency', (): void => {
      it('returns the invoice when found', async (): Promise<void> => {
        expect.assertions(1);

        const mockContext = makeMockContext();
        const invoiceRepository = makeMockInvoiceRepository();
        const mockInvoice = makeMockInvoice();

        (
          invoiceRepository.findById as jest.MockedFunction<any>
        ).mockResolvedValue(mockInvoice);

        const invoice = await invoiceService.retrieveInvoice(
          mockContext,
          invoiceRepository,
          mockInvoice.id,
        );

        expect(invoice).toStrictEqual(mockInvoice);
      });
    });
  });

  describe('retrieveInvoiceParties', (): void => {
    describe('error handling', (): void => {
      it('throws an error when the invoice repository throws an error', async (): Promise<void> => {
        expect.assertions(1);

        const mockContext = makeMockContext();
        const invoiceRepository = makeMockInvoiceRepository();
        const companyRepository = makeMockCompanyRepository();
        const errorToThrow = new Error('test error');

        (
          invoiceRepository.findById as jest.MockedFunction<any>
        ).mockRejectedValueOnce(errorToThrow);

        try {
          await invoiceService.retrieveInvoiceParties(
            mockContext,
            invoiceRepository,
            companyRepository,
            '123',
          );
        } catch (error: unknown) {
          expect(error).toEqual(errorToThrow);
        }
      });

      it('throws an error when the company repository throws an error', async (): Promise<void> => {
        expect.assertions(1);

        const mockContext = makeMockContext();
        const invoiceRepository = makeMockInvoiceRepository();
        const companyRepository = makeMockCompanyRepository();
        const errorToThrow = new Error('test error');

        (
          invoiceRepository.findById as jest.MockedFunction<any>
        ).mockResolvedValue(makeMockInvoice());
        (
          companyRepository.findById as jest.MockedFunction<any>
        ).mockRejectedValueOnce(errorToThrow);

        try {
          await invoiceService.retrieveInvoiceParties(
            mockContext,
            invoiceRepository,
            companyRepository,
            '123',
          );
        } catch (error: unknown) {
          expect(error).toEqual(errorToThrow);
        }
      });
    });

    describe('data persistency', (): void => {
      it('returns the issuer and recipient successfully', async (): Promise<void> => {
        expect.assertions(1);

        const mockContext = makeMockContext();
        const invoiceRepository = makeMockInvoiceRepository();
        const companyRepository = makeMockCompanyRepository();
        const mockInvoice = makeMockInvoice();
        const mockCompanyA = makeMockCompany({
          id: 'A',
        });
        const mockCompanyB = makeMockCompany({
          id: 'B',
        });

        (
          invoiceRepository.findById as jest.MockedFunction<any>
        ).mockResolvedValue(mockInvoice);
        (companyRepository.findById as jest.MockedFunction<any>)
          .mockResolvedValueOnce(mockCompanyA)
          .mockResolvedValueOnce(mockCompanyB);

        const parties = await invoiceService.retrieveInvoiceParties(
          mockContext,
          invoiceRepository,
          companyRepository,
          mockInvoice.id,
        );

        expect(parties).toMatchSnapshot();
      });
    });
  });
});
