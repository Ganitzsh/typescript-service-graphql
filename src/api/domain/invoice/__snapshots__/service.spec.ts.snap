// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invoice service listInvoices data persistency returns a page of invoices 1`] = `
{
  "cursor": undefined,
  "data": [
    {
      "currency": "EUR",
      "finalized": true,
      "id": "1",
      "issuer": "1",
      "phases": [
        "1",
        "2",
      ],
      "recipient": "2",
      "subtotal": 21,
      "total": 42,
    },
    {
      "currency": "EUR",
      "finalized": true,
      "id": "2",
      "issuer": "1",
      "phases": [
        "1",
        "2",
      ],
      "recipient": "2",
      "subtotal": 21,
      "total": 42,
    },
    {
      "currency": "EUR",
      "finalized": true,
      "id": "3",
      "issuer": "1",
      "phases": [
        "1",
        "2",
      ],
      "recipient": "2",
      "subtotal": 21,
      "total": 42,
    },
  ],
  "total": 3,
}
`;

exports[`invoice service retrieveInvoiceParties data persistency returns the issuer and recipient successfully 1`] = `
{
  "issuer": {
    "address": {
      "addressLine1": "test address line 1",
      "addressLine2": "test address line 2",
      "city": "test city",
      "country": "test country",
      "zipCode": "test zip code",
    },
    "id": "A",
    "name": "test company",
    "taxNumber": "test tax id",
  },
  "recipient": {
    "address": {
      "addressLine1": "test address line 1",
      "addressLine2": "test address line 2",
      "city": "test city",
      "country": "test country",
      "zipCode": "test zip code",
    },
    "id": "B",
    "name": "test company",
    "taxNumber": "test tax id",
  },
}
`;
